src/main/java/de/tum/in/www1/artemis/web/rest/NotificationResource.java

```diff
@@ -91,6 +103,30 @@ public NotificationResource(NotificationRepository notificationRepository, UserR
         return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
     }
 
+    /**
+     * GET /notifications/filtered-by-settings : Get all notifications that align with the current user notification settings by pages.
+     *
+     * @param pageable Pagination information for fetching the notifications
+     * @return the filtered list of notifications based on current user settings
+     */
+    @GetMapping("/notifications/filtered-by-settings")
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<List<Notification>> getAllNotificationsForCurrentUserFilteredBySettings(@ApiParam Pageable pageable) {
+        User currentUser = userRepository.getUserWithGroupsAndAuthorities();
+        Set<NotificationOption> notificationOptions = notificationOptionRepository.findAllNotificationOptionsForRecipientWithId(currentUser.getId());
+        Set<NotificationType> deactivatedTypes = notificationSettingsService.findDeactivatedNotificationTypes(notificationOptions);
+        Set<String> deactivatedTitles = notificationSettingsService.convertNotificationTypesToTitles(deactivatedTypes);
+        final Page<Notification> page;
+        if (deactivatedTitles.isEmpty()) {
+            page = notificationRepository.findAllNotificationsForRecipientWithLogin(currentUser.getGroups(), currentUser.getLogin(), pageable);
+        }
+        else {
+            page = notificationRepository.findAllNotificationsFilteredBySettingsForRecipientWithLogin(currentUser.getGroups(), currentUser.getLogin(), deactivatedTitles, pageable);
+        }
+        HttpHeaders headers = PaginationUtil.generatePaginationHttpHeaders(ServletUriComponentsBuilder.fromCurrentRequest(), page);
+        return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
+    }
+
```

Comment:
So, because it got requested, here is my take on it:

The shown solution has the benefit that we can always retrieve all notifications for the current user or filter them based on the settings.

Now we also have a proposed unified solution (that in this case results in the new method replacing the old, unless I misread something).

The real question to decide this is: what do we need? If the general one is not used anywhere, we can and should replace it. Adding it again in the future is easy, should we need it. If it is used in addition to the new one, we must keep it in some way.

I would, however, also like to provide two new perspectives.

We could enable the filter with an (optional) flag, therefore combining both methods into one. Should work fairly well here because of the similar code and functionality.

One step further would me my preferred way in terms of API design, provided that those options are pretty simple and short. It might however not be that much of an improvement as long as our own client is (pretty much) the only one. This would involve passing the filter options as part of the request, allowing the client to specify what it wants to receive. An empty set for example would just mean _all_ then.
As long as our client is the only supported one and we always filter based on user settings, this would be over-engineered.

So, it depends on what we actually need and use it for. I don't know the details and the context here right now.

One thing however is the REST endpoint. I think, unless we keep both, the endpoint should be called `.../notifications`, because there is no other option and notifications are then always filtered by current user and user settings. Or by additional, optional filter request parameters.

Reply:
Might also be interesting to know what the rest of the team / Stephan think about that.

Reply:
One thing to keep in mind in general for such situations, though: the git history will already keep it forever for you.

In the past, I was sometimes reluctant to remove code because it might be useful and used in the future, but in the end it will not be lost anyway and if it stays in the current code base, it was just a burden to work with and might not even get tested. I just deleted it and never had any regrets.

We can be especially brutal here in Artemis because we don't need to care about compatibility (yet).

Reply:
Thank you both very much for your deep dive into this issue. I appreciate your different views and ideas.
After thinking about what you have written I decided to remove the old implementation for receiving all notifications because I could not find a (current) use case where all notifications are used (regardless of settings). And I support your views that it is a good choice to remove things that are not used (we can easily recover/reintroduce it if need be), To leave unused code is an antipattern as we all know (lavaflow).
Thank you 👍